src/main/java/me/rayzr522/jsonmessage/compat/CompatManager.java

```diff
@@ -0,0 +1,74 @@
+package me.rayzr522.jsonmessage.compat;
+
+import me.rayzr522.jsonmessage.ReflectionHelper;
+import me.rayzr522.jsonmessage.compat.impl.*;
+
+public class CompatManager {
+    private static CompatManager INSTANCE;
+
+    private final ChatComponentCompat chatComponentCompat;
+    private final ChatPacketCompat chatPacketCompat;
+    private final PlayerConnectionCompat playerConnectionCompat;
+    private final TitlePacketCompat titlePacketCompat;
+
+    public CompatManager() {
+        int version = ReflectionHelper.getVersion();
+
+        chatComponentCompat = new ImplementationPicker<ChatComponentCompat>()
+                .addImplementation(8, 16, ChatComponentImpl8To16::new)
+                .addImplementation(17, Integer.MAX_VALUE, ChatComponentImpl17ToFuture::new)
+                .getImplementation(version)
+                .orElseThrow(() -> new IllegalStateException(
+                        "Missing ChatComponent implementation for major version: " + version
+                ));
+
+        chatPacketCompat = new ImplementationPicker<ChatPacketCompat>()
+                .addImplementation(8, 11, ChatPacketImpl8To11::new)
+                .addImplementation(12, 15, ChatPacketImpl12To15::new)
+                .addImplementation(16, 16, ChatPacketImpl16::new)
+                .addImplementation(17, Integer.MAX_VALUE, ChatPacketImpl17ToFuture::new)
+                .getImplementation(version)
+                .orElseThrow(() -> new IllegalStateException(
+                        "Missing ChatPacket implementation for major version: " + version
+                ));
+
+        playerConnectionCompat = new ImplementationPicker<PlayerConnectionCompat>()
+                .addImplementation(8, 16, PlayerConnectionImpl8To16::new)
+                .addImplementation(17, Integer.MAX_VALUE, PlayerConnectImpl17ToFuture::new)
+                .getImplementation(version)
+                .orElseThrow(() -> new IllegalStateException(
+                        "Missing PlayerConnection implementation for major version: " + version
+                ));
+
+        titlePacketCompat = new ImplementationPicker<TitlePacketCompat>()
+                .addImplementation(8, 16, TitlePacketImpl8To16::new)
+                .addImplementation(17, Integer.MAX_VALUE, TitlePacketImpl17ToFuture::new)
+                .getImplementation(version)
+                .orElseThrow(() -> new IllegalStateException(
+                        "Missing TitlePacket implementation for major version: " + version
+                ));
+    }
+
+    private static CompatManager get() {
+        if (INSTANCE == null) {
```

Comment:
main reason i do this is because failures in static initialization tend to result in failures that are a lot harder to control, i.e. if i am a consumer and i don't want my whole plugin to go kaboom when i try and use `JSONMessage` on an unsupported version then it's a lot easier to do that if the error gets thrown in a direct caller of the class rather than when the class just gets loaded